client.go

```diff
@@ -87,13 +77,37 @@ func (c *Client) NewSession(ctx context.Context, opts *SessionOptions) (*Session
 		HandleMax:      s.handleMax,
 	}
 	log.Debug(1, "TX (NewSession): %s", begin)
-	_ = s.txFrame(begin, nil)
+
+	// we use send to have positive confirmation on transmission
+	send := make(chan encoding.DeliveryState)
+	_ = s.txFrame(begin, send)
 
 	// wait for response
 	var fr frames.Frame
 	select {
 	case <-ctx.Done():
-		// TODO: this will leak s
+		select {
+		case <-send:
+			// begin was written to the network.  assume it was
+			// received and that the ctx was too short to wait for
+			// the ack. in this case we must send an end before we
+			// can delete the session
+			go func() {
```

Comment:
Frankly, these shenanigans make me a bit nervous.  The simple thing to do would be to kill `conn` but that seems a bit heavy-handed.